syntax = "proto3";
package protos.playbooks;

import "google/protobuf/wrappers.proto";

message Github {
  message FetchRelatedCommits {
    google.protobuf.StringValue owner = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue file_path = 3;
    google.protobuf.StringValue function_name = 4;
    google.protobuf.StringValue branch = 5;
  }

  message FetchFile {
    google.protobuf.StringValue owner = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue file_path = 3;
  }

  message UpdateFile {
    google.protobuf.StringValue owner = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue file_path = 3;
    google.protobuf.StringValue committer_name = 4;
    google.protobuf.StringValue committer_email = 5;
    google.protobuf.StringValue content = 6;
    google.protobuf.StringValue message = 7;
    google.protobuf.StringValue sha = 8;
    google.protobuf.StringValue branch = 9;
  }

  message FetchRecentCommits {
    google.protobuf.StringValue owner = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue branch = 3;
    google.protobuf.StringValue author = 4;
  }

  message FetchRecentMerges {
    google.protobuf.StringValue owner = 1;
    google.protobuf.StringValue repo = 2;
    google.protobuf.StringValue branch = 3;
  }

  message FileUpdate {
    google.protobuf.StringValue path = 1;
    google.protobuf.StringValue content = 2;
  }

  message CreatePullRequest {
    google.protobuf.StringValue repo = 1;
    google.protobuf.StringValue title = 2;
    google.protobuf.StringValue body = 3;
    google.protobuf.StringValue head_branch = 4;
    google.protobuf.StringValue base_branch = 5;
    repeated FileUpdate files = 6;
    google.protobuf.StringValue commit_message = 7;
    google.protobuf.StringValue committer_name = 8;
    google.protobuf.StringValue committer_email = 9;
  }

  enum TaskType {
    UNKNOWN = 0;
    FETCH_RELATED_COMMITS = 1;
    FETCH_FILE = 2;
    UPDATE_FILE = 3;
    FETCH_RECENT_COMMITS = 4;
    FETCH_RECENT_MERGES = 5;
    CREATE_PULL_REQUEST = 6;
  }

  TaskType type = 1;
  oneof task {
    FetchRelatedCommits fetch_related_commits = 2;
    FetchFile fetch_file = 3;
    UpdateFile update_file = 4;
    FetchRecentCommits fetch_recent_commits = 5;
    FetchRecentMerges fetch_recent_merges = 6;
    CreatePullRequest create_pull_request = 7;
  }
}